### PR TITLE
feat(sdk): add frontend build process to sdk

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David Adams <david@umaproject.org>

**Motivation**

https://app.shortcut.com/uma-project/story/2271/prepare-sdk-for-use-on-a-fe

**Summary**

This uses rollup-replace to go through and replace instances of `@uma/contracts-node` with `@uma/contracts-frontend` when building the frontend version. There ends up being 2 versions of the sdk in `dist/web` and `dist/node`.  Importing remains the same, so this seems to still run correctly on both api and a test frontend app without any changes.

This adds 2 new build commands
- `yarn build:web` - which builds to the frontend
- `yarn build:node` - which builds to node
- `yarn build` - which has been modified to run a script `build.js` so both builds are moved to independent folders

**Details**

Shows sdk size in frontend
image

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

https://app.shortcut.com/uma-project/story/2271/prepare-sdk-for-use-on-a-fe

### CODE REVIEW

File: packages/api/src/index.ts [hunk 1]
```diff
@@ -30,7 +30,7 @@ export type Actions = { [key: string]: Action };
 // this represents valid currencies to check prices against on coingecko
 // see: https://www.coingecko.com/api/documentations/v3#/asset_platforms/get_asset_platforms
 export type CurrencySymbol = "usd";
-export type PriceSample = [timestamp: number, price: string];
```
Comment (mrice32) [inline packages/api/src/index.ts:33]: Why drop the names?

File: packages/sdk/tsdx.config.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+module.exports = {
+  rollup(config) {
+    // removes the esm prefix -> that just makes importing the module more difficult
+    config.output.file = config.output.file.replace('.esm', '');
+
+    if (envType == 'web')
+      // unshift so that this plugin gets run first
+      config.plugins.unshift(
+        replace({ 
+          '@uma/contracts-node': "@uma/contracts-frontend",
```
Comment (mrice32) [inline packages/sdk/tsdx.config.js:13]: This is cool. Just to double check, is a string replace the standard way to replace imports? Also, does this replace it in the compiled output or does it modify the ts _before_ compilation? I assume the latter right?

File: packages/sdk/build.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
+// this file is adapted from https://blog.unterholzer.dev/cross-compatible-typescript-libraries/
```
Comment (mrice32) [inline packages/sdk/build.js:1]: Out of curiousity, why is this necessary -- shouldn't rollup create two separate builds for us and we should just need to point module and main at the different ones?

File: packages/sdk/src/index.ts [hunk 4]
```diff
@@ -8,11 +8,11 @@ export { default as Multicall2 } from "./multicall2";
 
 // types
 import type { TypedEventFilterEthers as TypedEventFilter, TypedEventEthers as TypedEvent } from "@uma/contracts-node";
-import { Contract } from "ethers";
-import { Result } from "@ethersproject/abi";
```
Comment (mrice32) [inline packages/sdk/src/index.ts:12]: Does removing these impact the bundle sizes?

**Review by mrice32 — APPROVED** [review]
Looks great! Only minor questions.

**Review by amateima — APPROVED** [review]

File: packages/api/src/index.ts (hunk 1 above)
Comment (daywiss) [reply packages/api/src/index.ts:33]: i was getting a build error for some reason. going back to see if i still get it

File: packages/sdk/tsdx.config.js (hunk 2 above)
Comment (daywiss) [reply packages/sdk/tsdx.config.js:13]: i didnt find any sources to show how to change imports based on environment during the build process. I kind of just tried this based on a slightly different tutorial for changing out code based on environment.  So no i dont think this is standard, but was not able to find a standard way to do this. 

I specifically had to push this plugin to the front of the stack so it ran before build. If you push this to the back the plugin array the build will fail!

File: packages/sdk/build.js (hunk 3 above)
Comment (daywiss) [reply packages/sdk/build.js:1]: This was copied from a tutorial. The problem is the sdk is using tsdx, which outputs the build to a single folder and will override the results of a previous build.  So this hacks around this by moving files around so previous build does not get overrwritten. There may be a better way to do this though

File: packages/sdk/src/index.ts (hunk 4 above)
Comment (daywiss) [reply packages/sdk/src/index.ts:12]: no it doesnt, this was an artifact of me trying to remove the ethersproject dependency.  I will leave it in since i found that you can access these things from the `ethers` library itself.

File: packages/sdk/build.js (hunk 3 above)
Comment (daywiss) [reply packages/sdk/build.js:1]: oh this file also copies the types file so that both builds have the same interface

File: packages/api/src/index.ts (hunk 1 above)
Comment (daywiss) [reply packages/api/src/index.ts]: looks like it does build with this in now
